fl4health/model_bases/fenda_base.py

```diff
@@ -1,58 +1,43 @@
-    CONCATENATE = "CONCATENATE"
-    SUM = "SUM"
-
-
-class FendaHeadModule(nn.Module, ABC):
-    def __init__(self, mode: FendaJoinMode) -> None:
-        super().__init__()
-        self.mode = mode
-
-    @abstractmethod
-    def local_global_concat(self, local_tensor: torch.Tensor, global_tensor: torch.Tensor) -> torch.Tensor:
-        raise NotImplementedError
-
-    @abstractmethod
-    def head_forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
-        raise NotImplementedError
-
-    def forward(self, local_tensor: torch.Tensor, global_tensor: torch.Tensor) -> torch.Tensor:
-        head_input = (
-            self.local_global_concat(local_tensor, global_tensor)
-            if self.mode == FendaJoinMode.CONCATENATE
-            else torch.add(local_tensor, global_tensor)
+class FendaModel(PartialLayerExchangeModel, ParallelSplitModel):
+    def __init__(self, local_module: nn.Module, global_module: nn.Module, model_head: ParallelSplitHeadModule) -> None:
+        ParallelSplitModel.__init__(
+            self, first_module=local_module, second_module=global_module, model_head=model_head
         )
-        return self.head_forward(head_input)
 
+    def layers_to_exchange(self) -> List[str]:
+        return [layer_name for layer_name in self.state_dict().keys() if layer_name.startswith("second_module.")]
 
-class FendaModel(PartialLayerExchangeModel):
-    def __init__(self, local_module: nn.Module, global_module: nn.Module, model_head: FendaHeadModule) -> None:
-        super().__init__()
-        self.local_module = local_module
-        self.global_module = global_module
-        self.model_head = model_head
 
-    def layers_to_exchange(self) -> List[str]:
-        return [layer_name for layer_name in self.state_dict().keys() if layer_name.startswith("global_module.")]
+class FendaModelWithFeatureState(FendaModel):
+    def __init__(
+        self,
+        local_module: nn.Module,
+        global_module: nn.Module,
+        model_head: ParallelSplitHeadModule,
+        flatten_features: bool = False,
+    ) -> None:
+        super().__init__(local_module, global_module, model_head)
+        self.flatten_features = flatten_features
 
     def forward(self, input: torch.Tensor) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
         # input is expected to be of shape (batch_size, *)
-        local_output = self.local_module.forward(input)
-        global_output = self.global_module.forward(input)
+        local_output = self.first_module.forward(input)
+        global_output = self.second_module.forward(input)
         preds = {"prediction": self.model_head.forward(local_output, global_output)}
         features = {
-            "local_features": local_output.reshape(len(local_output), -1),
-            "global_features": global_output.reshape(len(global_output), -1),
+            "local_features": local_output.reshape(len(local_output), -1) if self.flatten_features else local_output,
```

Comment:
This is a bit hard to read, I suggest you define the dict and then add each key on subsequent lines. Or some other way of breaking this up

Reply:
Will take a shot at making it a bit more readable 🙂